### PR TITLE
You can now put items into bags, belts and the suit slot using hotkeys.

### PR DESCRIPTION
**Pull request by Bm0n**
## What Does This PR Do

As the title says, you can now put items into belts, bags and the suit slot using hotkeys.

Default bindings are Shift E for belt Shift B for bag and Shift Q for suit slot. These can be changed in the hotkeys menu.

Semi-based on Cits implementation but the original code is from TG and thus is credited.

## Why It's Good For The Game

yeah i think this is okay for putting stuff into things yeah woo
## Testing
I messed around in-game to make sure nothing runtimed(don't quote me on that)

## Changelog
:cl: Bmon, ninjanomnom
add: You can now put items into your belts, bags and the suit slot using new hotkeys! (shiftE for belt, shiftB for bag, shiftQ for suit slot)
/:cl:

### CODE REVIEW

**Comment by AffectedArc07** [comment]
God damn I can smell the sweat from here

**Comment by Contrabang** [comment]
I like the quality of life of specifically storing things, but I think having a bind of being able to take out something in particular will get sweatly (e.g. stunbaton bind for belts).

**Comment by Bm0n** [comment]
> I like the quality of life of specifically storing things, but I think having a bind of being able to take out something in particular will get sweaty (e.g. stunbaton bind for belts).

I can definitely see the concern there which is why I am recommending this get TMed and not just merged out of the blue. I really want to see how this plays out on the server and see if it's as bad as it might be.

The QOL benefits are really worth it in my opinion and I don't think it'd change that much but yeah, TM to be safe.

**Comment by procdrone** [comment]
thats a bad idea.

**Comment by Qwertytoforty** [comment]
I am interested, will see how the tm goes

**Comment by matttheficus** [comment]
Belt and Suit Slot sounds fine. I dont like the idea of hot-grabbing items from Bags though, especially with how duffels work with zippers now. I dont want someone to hot-swap a Chainsaw out of a BOH.

**Comment by Bm0n** [comment]
> Belt and Suit Slot sounds fine. I dont like the idea of hot-grabbing items from Bags though, especially with how duffels work with zippers now. I dont want someone to hot-swap a Chainsaw out of a BOH.

If you wanted to draw stuff from your back slot quickly you probably shouldn't be wearing a duffel. I think that's a very poor argument, just don't wear a duffel or unzip it. It's kind of the whole point of the duffel, to act as long-term cold storage.

I think this issue lies mainly with an item rather than having the ability to draw without navigating UI.

SS13 is clunky and I knew that going into this PR the people who would be mainly against it are those who were accustomed to it. I can assure you on Paradise I can do anything that anyone else could do with Shift B or Shift E just by clicking and I am sure you can too!

The tick rate and movement here are so slow it doesn't really matter, it only penalized new players who are anxious and can't land their clicks on their items fast enough.

The Chainsaw is a bullshit item, it's been in my crosshairs to be toned back for a while now but this doesn't make it any better than it was before. Arguably if someone were using it it'd be out and in their hands to take advantage of its nodrop not tucked away in their BoH, but let's say it was being used in an ambush, it takes two actions to use it as a weapon. Pull, activate.

I think the other issue item would be the baton but I have an idea about that one and I might make accompanying PR here in a bit.

Still, the fact that we are talking about the balance of items shows the real issue to me.

**Comment by procdrone** [comment]
i apologize for my last message, it has no merit, no argument. ill explain a little.

Its a bad idea in my view since it promotes, and further increases "skill requirements" to combat, which moves the server further towards "action" side of play, while moving it away from roleplay side, and in my personal opinion, its not a good direction for the server. While reworking clunky combat systems to be more engaging is desirable, i just myself dont believe this to be the way.

**Comment by Bm0n** [comment]
> i apologize for my last message, it has no merit, no argument. ill explain a little.
> 
> Its a bad idea in my view since it promotes, and further increases "skill requirements" to combat, which moves the server further towards "action" side of play, while moving it away from roleplay side, and in my personal opinion, its not a good direction for the server. While reworking clunky combat systems to be more engaging is desirable, i just myself dont believe this to be the way.

I argue it does the exact opposite, levelling the playing field for all as now it is more assessable than flicking and as I explained above shouldn't really change anything for players that are already used to click. Hell, I still click on servers with a system like this all the time.

This has nothing to do with roleplay in the slightest, I have no idea why smoothing player to UI interactions would even bring this up. You could make this argument for any PR that touches combat or balance in any way.

We should never balance around having a bad player to UI interaction, to me that's bonkers. Instead, problem items should be the target of your frustrations.

**Comment by Exavere** [comment]
I was there on /tg/ before and when the hotkey mode was implemented, and I have been there for all the Combat Mode changes to date.  I'll say the PRs were met there with similar strife, and that yes this is a move away from old old tradition, but nowadays the Stamina Combat just wouldn't be the same without it.  The streamlined feel of combat just feels great.  

People who are already fairly robust will be able to enjoy a more in-pace fight, while those who were struggling to click their Baton in the belt before being DEsworded to death will find that it's nice to have a chance to fight before being decapitated.  

Not to mention it's just useful to have quick access to things during jobs.  Want a Medical Hypospray 5u Epi injection ASAP?  Shift+Q, inject and Shift+Q back it goes.  All while you simultaneously are reaching for the automender in your belt, if you aren't already Shift+E'ing for it -- Just my two cents with how this PR may go.  There will be changes to many things, a TM really will help sort out to what extent that is and if it's something that should be kept -- but it really is a nice QoL thing to have, especially with Stamina Combat.

**Comment by Bm0n** [comment]
> This isn't leveling the playing field, it's just giving those who have experience in servers with hotkeys an unfair and arbitrary advantage over Paradise veterans.

None of this is even arbitrary, you can rebind the keys to whatever you'd like.

I had to learn hotkeys, took me a few rounds tops. If I can do I am sure you can. It's not as big of a learning curve as you think it is, it's considerably easier than flicking.

> Hotkeys are not objectively more accessible than just clicking on stuff.

Why do you think this is the case? Please explain further.

> Cool story. We aren't /tg/ though.

Can you stop being rude to people for literally no reason just because they voice their opinions? Exa is someone who is not only a long player and administrator on TG but also here on Paradise.

It is important to consider what happened when a similar system was implemented on another server and his insight is helpful.

**Comment by hal9000PR** [comment]
I think we should all remain respectful in PR discussions please.
Also god damn bmon you sure love controversy.

I'm requesting a TM on this one. I'm open to the idea however i can see many peoples concerns.

**Comment by Bm0n** [comment]
>Also god damn bmon you sure love controversy.

wish i got some form of kick out of doing this, honestly hate it

but someone got to do it, plus I have the foresight of having interacted with this system before

**Comment by AffectedArc07** [comment]
> Let's just bring macros back and everyone will be happy.

No, never

**Comment by Bm0n** [comment]
> Let's just bring macros back and everyone will be happy.

Hilarious that you even bring this up considering it's completely unrelated.

Let me guess, your attitude here has something to do with me person who disabled byond macros, does it?

Regardless, I'll still treat your comments as serious, but half of this reads like concern trolling.

> Sure, I can "learn" hotkeys and become adequately proficient at using them.
> 
> The issue is, I'll have to deal with /tg/ sweatlords who have been playing around with hotkeys for months or even years. Do I have to play on /tg/ now just so I can get ready for future changes? Almost feels like I'm being punished for sticking to Paradise...

This again is a poor argument, you could literally make it about shoves, taser removal, the baton rework, TG cult and so on.

But to alleviate some of your concerns the crossover of players from TG to Paradise is very low, yes I am one of them but I rarely play and mostly just ghost and code. The other two people that I know of also rarely play, Exa being one of them.

I do find it a bit funny though that you have the opinion that everyone on TG is a sweatlord, you couldn't be any more wrong. More of a thing on Paradise than it is on TG sadly, an opinion of mine of course.

> Personal experience, but I still type emotes using Say intead of binding a hotkey to each one because it's just more convenient than having to remember 10 different hotkeys. Also subjectively, it just seems like a pain to press Shift+E/B/Q while running towards or away from someone (or even in general combat were you'll usually be moving around a lot). I only have so many fingers dedicated to the keyboard.

It's like you're ignoring what I am saying to you, I just told you in my previous message to you that you can rebind any of these keys to whatever you'd like. Shift E B Q is just the default.

But no one here is stopping you from clicking to your heart's content, you honestly won't lose much of an edge in combat if you so chose to and become skilled in flicking.

Still didn't really answer my question other than saying personal experience but 🤷 

> There wasn't a nicer way to say "this is not relevant to the discussion, your experiences on another server do not apply to Paradise".

You need to respect others opinions even if you don't agree with them. Yes, debate them, but don't outright dismiss them unless they are unwilling to listen back to you. It comes off as very rude.

If you have nothing constructive to add then don't say anything.

> All I see is "a similar system was implemented in a server almost completely distinct from Paradise in mechanics and culture and it was fine (_also, it's input from an admin of that server which means there's a potential conflict of interest_)". Can you explain how it's helpful?

Para is not as distinct as you think it is. Para and TG share a lot of systems and a very similar combat system. A combat system which is also shared with Yogs, Bee, Fulp, SR, Cit and many more servers. And guess what? All of them have hotkeys for drawing just like what I am trying to PR here.

Culture means nothing when we talk about balance and mechanics, although I wouldn't use the word culture.

Still, it would be foolish of me to PR this here and that be my only reason for wanting it ported. I have outlined why I think it is a good idea to port this and have at this point explained it to you. But to cut it short for you just as a reminder it meaningfully improves and smoothes out player to UI interactions without having to navigate a clunky menu and levels the playing field for all players.

Also to suggest someone who is an admin of both Paradise and TG has a conflict of interest is again very rude of you. I have no doubt in my mind that Exa is telling you his honest feelings about draw hotkeys. He saw this very feature right when it was added to TG, I didn't. His outlook on this is very insightful for not only me but everyone.

**Comment by Komrad822** [comment]
Just a question is the storing and taking out of items separate or are both just one key bind?

I.E: by default the keybind for bags for both is "shift+e" but I can have remap storing to "E" and leave taking out to "shift+e"

Good PR, just need this answered

**Comment by Bm0n** [comment]
> Just a question is the storing and taking out of items separate or are both just one key bind?
> 
> I.E: by default the keybind for bags for both is "shift+e" but I can have remap storing to "E" and leave taking out to "shift+e"
> 
> Good PR, just curious about this

It's combined into one proc. So if your hand is empty shift B will pull out the newest item from your bag and if your hand has an item in it shift B will store that item in the bag. The same goes for shift E and Q.

Technically possible to split the drawing and storing into two procs but having two proc bound to the same key wouldn't work and I assume the majority want it to just be one key for both, though I could be wrong.

**Comment by matttheficus** [comment]
> If you wanted to draw stuff from your back slot quickly you probably shouldn't be wearing a duffel. I think that's a very poor argument, just don't wear a duffel or unzip it. It's kind of the whole point of the duffel, to act as long-term cold storage.

This is not an argument for its use with duffel bags, this is an argument of not wanting people to be able to quick-draw from backpacks for balance reasons. Belts/Suit Slots, hell, pockets all are slots I would love to be able to quick-draw from. But, things like duffels, BoHs, and even normal bags allow for some silly things to be drawn from them. With this change I can holster a gun, suit slot a gun, and backpack a gun, and make it so I can lay down an impressive amount of firepower that would be hard pressed to defend from. 

> I can assure you on Paradise I can do anything that anyone else could do with Shift B or Shift E just by clicking and I am sure you can too!

I play a shit ton of MMOs, at a stupidly high level. I assure you I personally have zero issue with keybinding even the most mundane of things. I simply dislike the opportunities that this opens up for immense gamer'ing in a codebase that is NOT the same as others in terms of the balancing of Action and RP. No, I will not say "reeeeeeeeeeeee we're not TG/literally any TG downstream", but at the same time, please note that you cant just blanket "oh it works here, people got used to it there, so dont complain about it being difficult to adjust to".

> Why do you think this is the case? Please explain further.

This wasnt directed at us, but as someone who is an efficiency fiend in SS13 and MMOs, there is an unspoken "requirement" to learn to min-max to keep up with "high end" players. People like you, myself, and Rurik for examples would pick this up in 2-3 rounds tops and be right on our merry way singing praise for the QoL change. But, for the average player who has already committed the myriad of keybinds of SS13 to their memory, adding more stuff that theyll then have to use in stressful combat to keep up with the ever-lifting skill ceiling might have a bit more frustration. Changes in efficiency are exponentially better for "high skill" players and tend to make issues into a "rich get richer" problem.

I do personally agree that the QoL change outweighs this issue, BUT I can see the argument against it. Please keep in mind that some people find the addition of features that make combat more "streamlined" also tend to see it as an issue of Paradise moving more toward the "Action" side of the "Actions and RP" scale, including some of those in Leadership I've spoken to.

**Comment by S34NW** [comment]
This is a terrible idea, it will allow people to instantly smack people over the head with a weapon and no warning.
It will allow macros for doing such.

If you are doing this, it needs some downside like a delay before it is taken out

**Comment by S34NW** [comment]
An example of a bad consequence:

Someone writes an autohotkey macro that grabs your stun baton, clicks, then stores it. All within 1 second. You just put your mouse over the target and trigger your macro, the person is battoned with none of the expected counterplay where they can push you to steal it. You can repeat this macro as many times as you want to just win at combat.

This is a really bad idea in its current form.

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 1]
```diff
@@ -432,3 +432,88 @@
 /mob/living/carbon/human/proc/delete_equipment()
 	for(var/slot in get_all_slots())//order matters, dependant slots go first
 		qdel(slot)
+
+/mob/living/carbon/human/proc/quick_equip_bag() // take most recent item out of bag or place held item in bag
+	var/obj/item/thing = get_active_hand()
+	var/obj/item/storage/equipped_backpack = get_item_by_slot(slot_back)
+	if(ismecha(src.loc) || src.incapacitated(FALSE, TRUE))
+		return
```
Comment (hal9000PR) [inline code/modules/mob/living/carbon/human/inventory.dm:440]:
```suggestion
	if(ismecha(loc) || HAS_TRAIT(src, TRAIT_HANDS_BLOCKED))
		return
```
implied src and dont use `incapacitated`

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 2]
```diff
@@ -432,3 +432,88 @@
...
+	if(!stored || stored.on_found(src))
+		return
+	stored.attack_hand(src) // take out thing from backpack
+	return
+
+/mob/living/carbon/human/proc/quick_equip_belt() // take most recent item out of belt or place held item in belt
+	var/obj/item/thing = get_active_hand()
+	var/obj/item/storage/equipped_belt = get_item_by_slot(slot_belt)
+	if(ismecha(src.loc) || src.incapacitated(FALSE, TRUE))
+		return
```
Comment (hal9000PR) [inline code/modules/mob/living/carbon/human/inventory.dm:476]: same thing here

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 3]
```diff
@@ -432,3 +432,88 @@
...
+	if(!stored || stored.on_found(src))
+		return
+	stored.attack_hand(src) // take out thing from belt
+	return
+
+/mob/living/carbon/human/proc/quick_equip_suit() // takes things in and out of the suit slot
+	var/obj/item/thing = get_active_hand()
+	var/obj/item/equipped_suit = get_item_by_slot(slot_s_store)
+	if(ismecha(src.loc) || src.incapacitated(FALSE, TRUE))
+		return
```
Comment (hal9000PR) [inline code/modules/mob/living/carbon/human/inventory.dm:507]: and here

**Comment by Scribble-Sheep** [comment]
I'm not big on being able to take things out of stuff with a hotkey for...reasons I imagine have been already stated here, given the amount of comments I haven't bothered to read, but...I like having shortcuts for putting things in a specific slot. With the way E works at the moment it usually just...shoves it in my pocket, or stuffs my tool in my bag instead of my toolbelt.
Or god forbid, in some weird ass slot like behind my ear, causing me to look for it later and think it got stolen when I can't find it.

With specific stash keybinds alone, I don't think it'd be too bad of a power gamey thing, since we already have an auto-stash keybind anyway.

**Comment by TORE01** [comment]
While I love the idea of making combat even faster I'd much prefer if we could add to what we already have instead of adding a quick equip system from tg. We have shoulder holsters IG that allows you to holster and unholster guns very fast with one button. This works very well because despite being a powerful item both for sec and antags, it's locked behind botany and has good restriction when it comes to which weapons can be stored in it.
I would love it if a regular holster were added that can be used in a similar way. And while you're at it, make cowboy hats craftable too. Yeeeehaw.

**Comment by Bm0n** [comment]
> This is a terrible idea, it will allow people to instantly smack people over the head with a weapon and no warning. It will allow macros for doing such.
> 
> If you are doing this, it needs some downside like a delay before it is taken out
>
> An example of a bad consequence:
> 
> Someone writes an autohotkey macro that grabs their stun baton, clicks, then stores it. All within 1 second. They just put their mouse over the target and trigger their macro, the person is battoned with none of the expected counterplay where they can push back to steal it. They can repeat this macro as many times as they want to just win at combat.

You can do this right now with almost zero difference with AHK or AutoIt using emulated clicks. I am surprised you say this considering you have coding knowledge.

If someone had the intention to code something like this there would be no difference.

**Comment by Bm0n** [comment]
> This is not an argument for its use with duffel bags, this is an argument of not wanting people to be able to quick-draw from backpacks for balance reasons. Belts/Suit Slots, hell, pockets all are slots I would love to be able to quick-draw from. But, things like duffels, BoHs, and even normal bags allow for some silly things to be drawn from them. With this change I can holster a gun, suit slot a gun, and backpack a gun, and make it so I can lay down an impressive amount of firepower that would be hard pressed to defend from.

Again I understand the concerns and am compiling a list of problem items the chainsaw being one of them and will work on fixes for them.

Guns shouldn't be that big of an issue, we have done many things over the years to limit them like shitcanning sci guns. This is why I am curious to see what happens in a TM. I think a TM would put a lot of concerns to rest.

> I play a shit ton of MMOs, at a stupidly high level. I assure you I personally have zero issue with keybinding even the most mundane of things. I simply dislike the opportunities that this opens up for immense gamer'ing in a codebase that is NOT the same as others in terms of the balancing of Action and RP. No, I will not say "reeeeeeeeeeeee we're not TG/literally any TG downstream", but at the same time, please note that you cant just blanket "oh it works here, people got used to it there, so dont complain about it being difficult to adjust to".

It won't be as bad as you think it'd be. Again, this levels the playing field more than making it gamery and that's what I like about hotkeys. A great equalizer if you will. No longer will someone's skill be determined by how fast they can flick but by how well they control their nerves and land hits, it removes that skill gap if flicking. This isn't CSGO, I don't think anyone should be forced to flick their equippes/de-equippes.

> This wasnt directed at us, but as someone who is an efficiency fiend in SS13 and MMOs, there is an unspoken "requirement" to learn to min-max to keep up with "high end" players. People like you, myself, and Rurik for examples would pick this up in 2-3 rounds tops and be right on our merry way singing praise for the QoL change. But, for the average player who has already committed the myriad of keybinds of SS13 to their memory, adding more stuff that theyll then have to use in stressful combat to keep up with the ever-lifting skill ceiling might have a bit more frustration. Changes in efficiency are exponentially better for "high skill" players and tend to make issues into a "rich get richer" problem.

It wasn't but I'll still respond to you regardless as if it were.

I don't deny there will probably be a teething period for some players but ultimately this lowers the skill gap and levels things about way better than what we had before which again is flicking. Me and you can both probably flick our attacks very well but the average person cannot and it takes a considerably longer amount of time to train yourself to do that over learning shift + E.

I will say though the keybinds that you tend to use in high stress situations are usually the easiest to remember and keep in your memory. That's just how memory works and how we associate actions and things with memorable events.

> I do personally agree that the QoL change outweighs this issue, BUT I can see the argument against it. Please keep in mind that some people find the addition of features that make combat more "streamlined" also tend to see it as an issue of Paradise moving more toward the "Action" side of the "Actions and RP" scale, including some of those in Leadership I've spoken to.

I do not feel as if this impacts roleplay in any way. People will still choice to RP if they so wish to or not if they don't. This really won't change things, RP is more of an issue with the community rather than balance decisions.

**Comment by Mitchs98** [comment]
Honestly do not want to see this merged due to many of the same reasons described above. Even if you _can_ do this with scripting it doesn't mean most people do it at all, at the very least I entirely doubt that it's common practice. Nor is it something I believe should essentially be baked into the core game. 

I am aware that it's supposed to provide utility to other jobs...but I really don't see many use-cases for this, that aren't purely hypothetical and theoretical, for anything other than combat. Be it from antags, validhunters, security, or general powergamers. They're going to optimize where to put their weapons, meds, and whatever else will help them game harder.

This is, of course, assuming that it retains the functionality of being able to automatically/quickly pull the last item from a given inventory space with the same key.

Now, I'm all for this PR if it's simply putting things away. That's a good QOL change that is in-fact almost entirely QOL with little to no implications for using it to game combat. As it stands now being able to retrieve items too is just too far.

**Comment by S34NW** [comment]
>
> You can do this right now with almost zero difference with AHK or AutoIt using emulated clicks. I am surprised you say this considering you have coding knowledge.

Considering mouse movements are currently involved it would be a lot more technically challenging for people to do this, as-is.

The system being put forward here means no need to faff about with moving a mouse around, just two button presses with a click in the middle.

**Comment by Denthamos** [comment]
I have concerns with the combat implications of this and it feels like this will require item equip/use delay added to certain items due to this.  Taking a baton off a belt requiring a delay, taking a Cr-20 out of a backpack needing a delay, ect.  That CM style mentality was honestly annoying as hell and not a direction I want this server to go nor will I sign off on such.

**Comment by Rurik123** [comment]
I'm fairly neutral on this, but I do have one question:

I thought the reason we removed custom macros was to move away from the reliance on them, shifting back onto manually typing/clicking, thus leveling the playing field for 90% of the playerbase that didn't use pickup or callout macros to game. This PR leaves me a bit confused, with it taking us back to where we started. Exchanging pickup macros for auto draw from bag, belt, and suit storage seems like we just walked in a circle. 

> Considering mouse movements are currently involved it would be a lot more technically challenging for people to do this, as-is.
> 
> The system being put forward here means no need to faff about with moving a mouse around, just two button presses with a click in the middle.

Its the same on a technical level. If you can put together one, you can put together the other. _However_ this will make the process more reliable. Emulated clicks (as well as regular clicks) are sometimes fumbled by byond and not recognized, depending on time dialation. Hotkeys, on average, are not only quicker than an emulated click but also more reliable, rarely missed upon input.

**Comment by Bm0n** [comment]
> Considering mouse movements are currently involved it would be a lot more technically challenging for people to do this, as-is.
> 
> The system being put forward here means no need to faff about with moving a mouse around, just two button presses with a click in the middle.

You already have to use emulated clicks to spam attack it's not that different. you're targeting a specific location with emulated clicks then spam clicking at the cursor location for 0.5 of a second then clicking back to where you drew from

i know this because I have made autoit scripts before, not for ss13 thought but still you can do this very easily if you know how to do the other

this is actual fearmongering, no other good way to describe it, sorry sean

> Its the same on a technical level. If you can put together one, you can put together the other. _However_ this will make the process more reliable. Emulated clicks (as well as regular clicks) are sometimes fumbled by byond and not recognized, depending on time dialation. Hotkeys, on average, are not only quicker than an emulated click but also more reliable, rarely missed upon input.

I have never had one of my clicks ignored by byond. Only maybe if the server is lagging or there is a high TD but it always still eventually draws.  Maybe that's because I have good ping? I don't know.

**Comment by procdrone** [comment]
> > i apologize for my last message, it has no merit, no argument. ill explain a little.
> > Its a bad idea in my view since it promotes, and further increases "skill requirements" to combat, which moves the server further towards "action" side of play, while moving it away from roleplay side, and in my personal opinion, its not a good direction for the server. While reworking clunky combat systems to be more engaging is desirable, i just myself dont believe this to be the way.
> 
> I argue it does the exact opposite, levelling the playing field for all as now it is more assessable than flicking and as I explained above shouldn't really change anything for players that are already used to clicking. Hell, I still click on servers with a system like this all the time.
> 
> This has nothing to do with roleplay in the slightest, I have no idea why smoothing player to UI interactions would even bring this up. You could make this argument for any PR that touches combat or balance in any way.
> 
> We should never balance around having a bad player to UI interaction, to me that's bonkers. Instead, problem items should be the target of your frustrations.

I think you missed my meaning in this. Let me elaborate. Making actions that usually meant you had to open the relevant bag/belt and take the item out, then store it back. it was less convenient but I also believe it was more fair for everyone. there are more delays. meant sweats can't 360 no scope you spamming hotkeys. Of course there is still a lot of that (sweats winning) in current combat but in my view, this PR will only widen the gap.

Now, to engage with combat you WILL have to sweat yourself. This moves the server firmly forward to Action side of things, this by itself moves it away from RP. Paradise is supposed to be "a perfect mix of Action and RP", so you can't just heavily build on one element not hurting the other.

I for one, suffer from very often, void inputs. (i click on thing and nothing happens), same with already existing hotkeys. I often fail to switch a hand or take an item out of a bag despite clicking. And in combat these missed seconds already count. with your PR i wont even have time to correct that mistake/problem, since the other guy will be able to just hotkey his way through combat while im combating the UI.

To me, this will make combat less enjoyable, as it will rely on skill and sweating more, and this is not an experience i want from paradise. 

But of course, thats just my opinion, everyone is welcome to agree or disagree. Not trying to force anyone to accept my view, i just wish to share it since I like playing on the server.

**Comment by A-Medster** [comment]
From what ive seen, the quick storage part of the PR works well and is really nice to have
Though the quick draw aspect seems to be a little iffy if its working how i think it is.
Being able to quick draw from any three inventorys on your character seems like it could be gamed, best thing i can think of here would likely being the quick draw feature only applys to your most recently interacted with item and nothing else.
Same basic concept, just toned down from three different items

**Comment by S34NW** [comment]
https://user-images.githubusercontent.com/12197162/209315732-87309095-3fc0-42dc-ada3-ba034f95446c.mp4

So this is an example of the behaviour that worries me, I can do this even without a damn macro. No counterplay possible, no way to steal their weapon or shove them so they drop it.
This massively changes the current shovefight meta in what I think is a negative way. Sure this will be harder to do during a proper combat scenario, but with a little more muscle memory for the timing it would definitely be reasonable to pull off.

**Comment by Bm0n** [comment]
> ezgif.com-gif-maker.mp4 
> So this is an example of the behaviour that worries me, I can do this even without a damn macro. No counterplay possible, no way to steal their weapon or shove them so they drop it. This massively changes the current shovefight meta in what I think is a negative way. Sure this will be harder to do during a proper combat scenario, but with a little more muscle memory for the timing it would definitely be reasonable to pull off.

https://user-images.githubusercontent.com/92271472/209322145-297d6b65-cf33-4b0e-902c-c8d5b7f12cb3.mp4

Yes slower and I have been considerably balder as of late but your example is why I made PR #19981, to add a delay back into the mix.

Also, mind you if you were fighting someone that was moving all over the place you'd need to keep your baton out to try to hand a hit. You get a shove on them while they were trying to do that and they're done for.

not to mention TD and ping, oh god, no way realistically it'd be that fast.

**Comment by S34NW** [comment]
> not to mention TD and ping, oh god, no way realistically it'd be that fast.

My video was recorded on live while all manner of nonsense was going on

**Comment by Bm0n** [comment]
> My video was recorded on live while all manner of nonsense was going on (hence me not including the chat window)

hrrrrng its very low pop right now though. but yeah fair enough

**Comment by hal9000PR** [comment]
I am not sure the instant draw is gonna be healthy for the game in the long run. the putting things ***into*** storage however seem incredibly QOL and helpful

**Comment by Bm0n** [comment]
> I am not sure the instant draw is gonna be healthy for the game in the long run. the putting things _**into**_ storage however seem incredibly QOL and helpful

I am once again disappointed, this very feature has worked out very well in the long term on almost every other server that has ported it which is almost all of them, Goon probably being the most notable exception. Sad to see we're stuck in the stone age of UI interactions here but that's life, things almost never turn out the way you wanted them to.

good luck with combat mode Hal, you're going to need it considering the reaction this got.

drawing is gone

**Comment by AffectedArc07** [comment]
> things almost never turn out the way you want them to.

Roundstart blob moment

---

> Drawing is gone, title and changelog adjusted

Could you get the PR desc too please?

**Comment by Bm0n** [comment]
> > things almost never turn out the way you want them to.
> 
> Roundstart blob moment
> 
> > Drawing is gone, title and changelog adjusted
> 
> Could you get the PR desc too please?

yes

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 4]
```diff
@@ -432,3 +432,40 @@
...
+			return
+
+/mob/living/carbon/human/proc/quick_equip_suit() // puts things in the suitslot
+	var/obj/item/thing = get_active_hand()
+	var/obj/item/equipped_suit = get_item_by_slot(slot_s_store)
+	if(ismecha(loc) || HAS_TRAIT(src, TRAIT_HANDS_BLOCKED))
+		return
+	if(!equipped_suit)
+		equip_to_slot_if_possible(thing, slot_s_store)
+		return
```
Comment (Sirryan2002) [inline code/modules/mob/living/carbon/human/inventory.dm:471]:
```suggestion
```
dangling return

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 5]
```diff
@@ -432,3 +432,40 @@
...
+	if(ismecha(loc) || HAS_TRAIT(src, TRAIT_HANDS_BLOCKED))
+		return
+	if(!equipped_backpack) // We also let you equip a backpack like this
+		equip_to_slot_if_possible(thing, slot_back)
+		return
+	if(thing) // put thing in backpack
+		if(equipped_backpack.can_be_inserted(thing))
+			equipped_backpack.handle_item_insertion(thing)
+			playsound(loc, "rustle", 50, 1, -5)
+			return
```
Comment (Sirryan2002) [inline code/modules/mob/living/carbon/human/inventory.dm:448]:
```suggestion
```
dangling return

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 6]
```diff
@@ -432,3 +432,35 @@
...
+	var/obj/item/thing = get_active_hand()
+	var/obj/item/storage/equipped_belt = get_item_by_slot(slot_belt)
+	if(ismecha(loc) || HAS_TRAIT(src, TRAIT_HANDS_BLOCKED))
+		return
+	if(!istype(equipped_belt)) // We also let you equip a belt like this
+		equip_to_slot_if_possible(thing, slot_belt)
+		return
+	if(thing && equipped_belt.can_be_inserted(thing)) // put thing in belt
+		equipped_belt.handle_item_insertion(thing)
+		playsound(loc, "rustle", 50, 1, -5)
```
Comment (lewcc) [inline code/modules/mob/living/carbon/human/inventory.dm:458]: This code is somewhat duplicated. Think you could consolidate it to a general quick_equip proc and pass in the slot you want to equip things to?

Comment (Bm0n) [reply code/modules/mob/living/carbon/human/inventory.dm:458]: How would you be able to rebind it then? They all need to be a separate proc for each bind

Comment (S34NW) [reply code/modules/mob/living/carbon/human/inventory.dm:458]: Actually, you could just use the same proc but pass a different arg for the slot depending on the keybind

Comment (Bm0n) [reply code/modules/mob/living/carbon/human/inventory.dm]: should be better now, based on what Vi3trice recommended I should do

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 7]
```diff
@@ -432,3 +432,24 @@
...
+        playsound(loc, "rustle", 50, 1, -5)
+
+/mob/living/carbon/human/proc/quick_equip_bag()
+    quick_equip_item(slot_back)
+
+/mob/living/carbon/human/proc/quick_equip_belt()
+    quick_equip_item(slot_belt)
+
+/mob/living/carbon/human/proc/quick_equip_suit()
+	quick_equip_item(slot_s_store)
```
Comment (hal9000PR) [inline code/modules/mob/living/carbon/human/inventory.dm:455]: why not make the macros call `quick_equip_item` with different args?

File: code/datums/keybindings/human.dm [hunk 8]
```diff
@@ -4,6 +4,33 @@
...
+	M.quick_equip_belt()
+
+/datum/keybinding/human/suit_equip
+	name = "Equip Held Object To Suit Storage"
+	keys = list("ShiftQ")
+
+/datum/keybinding/human/suit_equip/down(client/C)
+	. = ..()
+	var/mob/living/carbon/human/M = C.mob
+	M.quick_equip_suit()
```
Comment (hal9000PR) [inline code/datums/keybindings/human.dm:32]:
```suggestion
/datum/keybinding/human/suit_equip/down(client/C)
	. = ..()
	var/mob/living/carbon/human/M = C.mob
	M.quick_equip_item(slot_s_store)
```
like this,  for all the macros.

**Review by hal9000PR — CHANGES_REQUESTED** [review]

File: code/datums/keybindings/human.dm (hunk 8 above)
Comment (Bm0n) [reply code/datums/keybindings/human.dm]: yeah that works i guess

**Review by hal9000PR — APPROVED** [review]

**Pull request merged**